src/map/enchantui.c

```diff
@@ -0,0 +1,930 @@
+/**
+ * This file is part of Hercules.
+ * http://herc.ws - http://github.com/HerculesWS/Hercules
+ *
+ * Copyright (C) 2021 Hercules Dev Team
+ *
+ * Hercules is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#define HERCULES_CORE
+
+#include "enchantui.h"
+
+#include "common/db.h"
+#include "common/memmgr.h"
+#include "common/mmo.h"
+#include "common/nullpo.h"
+#include "common/random.h"
+#include "common/showmsg.h"
+#include "common/timer.h"
+#include "map/battle.h"
+#include "map/chrif.h"
+#include "map/clif.h"
+#include "map/pc.h"
+#include "map/refine.h"
+
+#include <stdlib.h>
+
+static struct enchantui_interface enchantui_s;
+struct enchantui_interface *enchantui;
+
+static void enchantui_read_db_libconfig(void)
+{
+	char filepath[512];
+	snprintf(filepath, sizeof(filepath), "%s/%s", map->db_path, DBPATH"enchant_db.conf");
+
+	struct config_t enchant_conf;
+	if (libconfig->load_file(&enchant_conf, filepath) == CONFIG_FALSE) {
+		ShowError("%s: can't read %s\n", __func__, filepath);
+		return;
+	}
+
+	struct config_setting_t *enchant_db = NULL;
+	if ((enchant_db = libconfig->setting_get_member(enchant_conf.root, "enchant_db")) == NULL) {
+		ShowError("%s: can't read %s\n", __func__, filepath);
+		return;
+	}
+
+	int i = 0;
+	int count = 0;
+	struct config_setting_t *it = NULL;
+
+	while ((it = libconfig->setting_get_elem(enchant_db, i++)) != NULL) {
+		if (enchantui->read_db_libconfig_sub(it, i, filepath))
+			++count;
+	}
+
+	libconfig->destroy(&enchant_conf);
+	ShowStatus("Done reading '"CL_WHITE"%d"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", count, filepath);
+}
+
+static bool enchantui_read_db_libconfig_sub(const struct config_setting_t *it, int n, const char *source)
+{
+	nullpo_retr(false, it);
+	nullpo_retr(false, source);
+
+	struct enchant_info ei = { 0 };
+
+	if (libconfig->setting_lookup_int(it, "Id", &ei.Id) == CONFIG_FALSE) {
+		ShowError("%s: Invalid Enchant Id provided for entry %d in '%s', skipping...\n", __func__, n, source);
+		return false;
+	}
+
+	libconfig->setting_lookup_int(it, "MinRefine", (int *)&ei.MinRefine);
```

Comment:
`MinRefine` is an 8 bit integer, passing it as an int pointer to setting_lookup_int() may cause it to overwrite memory around it.

It would be better to use a separate int32 variable and then assign to MinRefine.